sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,850 @@
+                                    )
+
+                    self._call_back(
+                        CallbackType.CONNECTED,
+                        OnConnectedArgs(connection_id=message.connection_id, user_id=message.user_id),
+                    )
+
+            def handle_disconnected_message(message: DisconnectedMessage):
+                self._last_disconnected_message = message
+
+            def handle_group_data_message(message: GroupDataMessage):
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallbackType.GROUP_MESSAGE,
+                    OnGroupDataMessageArgs(
+                        data_type=message.data_type,
+                        data=message.data,
+                        group=message.group,
+                        from_user_id=message.from_user_id,
+                        sequence_id=message.sequence_id,
+                    ),
+                )
+
+            def handle_server_data_message(message: ServerDataMessage):
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallbackType.SERVER_MESSAGE,
+                    OnServerDataMessageArgs(
+                        data_type=message.data_type, data=message.data, sequence_id=message.sequence_id
+                    ),
+                )
+
+            parsed_message = self._protocol.parse_messages(data)
+            if parsed_message is None:
+                #  None means the message is not recognized.
+                return
+            if parsed_message.kind == "connected":
+                handle_connected_message(parsed_message)
+            elif parsed_message.kind == "disconnected":
+                handle_disconnected_message(parsed_message)
+            elif parsed_message.kind == "ack":
+                handle_ack_message(parsed_message)
+            elif parsed_message.kind == "groupData":
+                handle_group_data_message(parsed_message)
+            elif parsed_message.kind == "serverData":
+                handle_server_data_message(parsed_message)
+            else:
+                _LOGGER.warning("unknown message type: %s", parsed_message.kind)
+
+        def on_close(_: Any, close_status_code: int, close_msg: str):
+            if self._state == WebPubSubClientState.CONNECTED:
+                _LOGGER.info("WebSocket connection closed. Code: %s, Reason: %s", close_status_code, close_msg)
+
+                self._last_close_event = CloseEvent(close_status_code=close_status_code, close_reason=close_msg)
+                # clean ack cache
+                self._ack_map.clear()
```

Comment:
Do we need to notify before clear?

Reply:
good point! Although we have ack_timeout, it is better to notify before clear to unblock the process.
Updated